pkg/reconciler/tektonwrapper/tektonwrapper.go

```diff
@@ -0,0 +1,271 @@
+package tektonwrapper
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	quotav1 "github.com/openshift/api/quota/v1"
+	"github.com/redhat-appstudio/jvm-build-service/pkg/apis/jvmbuildservice/v1alpha1"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	//TODO eventually we'll need to decide if we want to make this tuneable
+	contextTimeout = 300 * time.Second
+
+	TektonWrapperId = "jvmbuildservice.io/tektonwrapperid"
+)
+
+var (
+	log = ctrl.Log.WithName("tektonwrapper")
+)
+
+type ReconcileTektonWrapper struct {
+	client           client.Client
+	nonCachingClient client.Client
+	scheme           *runtime.Scheme
+	eventRecorder    record.EventRecorder
+}
+
+func newReconciler(mgr ctrl.Manager, nonCachingClient client.Client) reconcile.Reconciler {
+	return &ReconcileTektonWrapper{
+		client:           mgr.GetClient(),
+		nonCachingClient: nonCachingClient,
+		scheme:           mgr.GetScheme(),
+		eventRecorder:    mgr.GetEventRecorderFor("TektonWrapper"),
+	}
+}
+
+func (r *ReconcileTektonWrapper) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	// Set the ctx to be Background, as the top-level context for incoming requests.
+	ctx, cancel := context.WithTimeout(ctx, contextTimeout)
+	defer cancel()
+	tw := v1alpha1.TektonWrapper{}
+	twerr := r.client.Get(ctx, request.NamespacedName, &tw)
+	if twerr != nil {
+		if !errors.IsNotFound(twerr) {
+			return ctrl.Result{}, twerr
+		}
+	}
+
+	pr := &v1beta1.PipelineRun{}
+	prerr := r.client.Get(ctx, request.NamespacedName, pr)
+	if prerr != nil {
+		if !errors.IsNotFound(prerr) {
+			return ctrl.Result{}, prerr
+		}
+	}
+
+	if prerr != nil && twerr != nil {
+		msg := "Reconcile key %s received not found errors for both pipelineruns and tektonwrapper (probably deleted)\"" + request.NamespacedName.String()
+		log.Info(msg)
+		return ctrl.Result{}, nil
+	}
+
+	switch {
+	case prerr == nil:
+		if pr.IsDone() {
+			twId, ok := pr.Labels[TektonWrapperId]
+			if !ok || len(twId) == 0 {
+				msg := fmt.Sprintf("pipelinerun %s unexpectedly missing tekton wrapper ID", request.NamespacedName.String())
+				r.eventRecorder.Event(pr, corev1.EventTypeWarning, "MissingTektonWrapperID", msg)
+				return reconcile.Result{}, nil
+			}
+			twKey := types.NamespacedName{
+				Namespace: pr.Namespace,
+				Name:      twId,
+			}
+			if err := r.client.Get(ctx, twKey, &tw); err != nil {
+				return reconcile.Result{}, err
+			}
+			tw.Status.State = v1alpha1.TektonWrapperStateComplete
```

Comment:
Can we look for a throttled pipeline here, and re-queue it somehow? Having a fixed wait time seems kinda inefficent and possibly prone to the thundering herd problem.

Reply:
ooh interesting; I'll look into that today, though I will probably push updates for the other comments first without this change; I'll be sure and announce in a comment when I have something along these lines

Reply:
ok I have something for this looking good locally; vetting a bit more, then will push an update

Reply:
OK I've pushed the update for this as well.  I think this change really put it over the top wrt the positive local run I reported in slack @stuartwdouglas

Reply:
see https://github.com/redhat-appstudio/jvm-build-service/pull/183/files#diff-669e675d95667806a524632fe78e880ab9dabe1839a85447b01bfd92b53fea99R104